### ISSUE TITLE
resetProcessMetrics leaves orphaned objects behind

### ISSUE THREAD

**Issue by kgriesh** [idanto/node-metrics#19]
Calls to resetMetric() does not clean up any interval objects that are running. This happens with gc.releasedMem and gc.time processes because they use Meter objects which spawn intervals, which are never cleared. The result is a memory leak, new orphaned Timer and Meter objects getting left on the heap with each call to resetProcessMetrics(). To help illustrate, this is a hack I added locally to confirm and it does fix it but not in an elegant way.

  if (trackedMetrics[NAMESPACES.process].gc) {
    trackedMetrics[NAMESPACES.process].gc.releasedMem._meter.end();
    trackedMetrics[NAMESPACES.process].gc.time._meter.end();
  }

**Comment by idanto** [comment]
@kgriesh Thanks for letting me know about the issue.
As you said this issue is relevant to all Meter and Timer objects (there are some more than just the process metrics).

My suggested solution is to add a function that will call `end()` for all Meter and Timer objects. I will call this function from `resetMetric()` and it will run recursively on all metrics:

```js
function resetMetric(namespaceToReset) {
  stopMeterMetrics(trackedMetrics[namespaceToReset]);
  delete trackedMetrics[namespaceToReset];
}

function stopMeterMetrics(metric) {
  for (var subMetricsNames in metric) {
    if (metric[subMetricsNames] instanceof measured.Meter) {
      metric[subMetricsNames].end();
    } else if (metric[subMetricsNames] instanceof measured.Timer) {
      metric[subMetricsNames]._meter.end();
    } else if ((metric[subMetricsNames] instanceof measured.Gauge) || (metric[subMetricsNames] instanceof measured.Counter)) {
      //Do nothing
    } else if (metric[subMetricsNames] instanceof Object) {
      stopMeterMetrics(metric[subMetricsNames]);
    }
  }
}
```

Can you let me know if it works for you?
if the answer is yes I have to think about a good way to test it automatically and then I will release a new version with the fix.

**Comment by kgriesh** [comment]
Yes. That would work for us. Thanks for the quick response.

**Comment by idanto** [comment]
released under version 1.4.2.

Thanks for your issue @kgriesh .
